js.go

```diff
@@ -1250,6 +1369,128 @@ func (jsi *jsSub) trackSequences(reply string) {
 	jsi.mu.Unlock()
 }
 
+// Check to make sure messages are arriving in order.
+// Returns true if the sub had to be replaced. Will cause upper layers to return.
+// Lock should be held.
+func (sub *Subscription) checkOrderedMsgs(m *Msg) bool {
+	// Ignore msgs with no reply like HBs and flowcontrol, they are handled elsewhere.
+	if m.Reply == _EMPTY_ || sub.jsi == nil || isControlMessage(m) {
+		return false
+	}
+
+	// Normal message here.
+	tokens, err := getMetadataFields(m.Reply)
+	if err != nil {
+		return false
+	}
+	sseq, dseq := uint64(parseNum(tokens[5])), uint64(parseNum(tokens[6]))
+
+	jsi := sub.jsi
+	jsi.mu.Lock()
+	if dseq != jsi.dseq {
+		jsi.mu.Unlock()
+		sub.resetOrderedConsumer(jsi.sseq + 1)
+		return true
+	}
+	// Update our tracking here.
+	jsi.dseq, jsi.sseq = dseq+1, sseq
+	jsi.mu.Unlock()
+	return false
+}
+
+// We are here if we have detected a gap with an ordered consumer.
+// We will create a new consumer and rewire the low level subscription.
+// Lock should be held.
+func (sub *Subscription) resetOrderedConsumer(sseq uint64) {
+	nc := sub.conn
+	closed := sub.closed
+	if sub.jsi == nil || nc == nil || closed {
+		return
+	}
+
+	// Quick unsubscribe. Since we know this is a simple push subscriber we do in place.
+	nc.subsMu.Lock()
```

Comment:
Another one: subsMu->sub.mu, not the opposite. Maybe make resetOrderedConsumer outside of the sub lock (and grab sub.mu lock above only to gather closed/etc..)

Reply:
I need to make sure no more messages are received after this one in the stack. I could release the sub lock here though.

Reply:
From my reading of the code, this is invoked from the readLoop isn't it? So messages would not be added to the sub, even if we release the lock.

Reply:
Will clean up.

Reply:
Yes just can't do it another go routine is what I meant.